src/freedv_interface.cpp

```diff
@@ -447,84 +459,184 @@ int FreeDVInterface::processRxAudio(
+                freq_shift_coh(rx_fdm_offset, rx_fdm, st->rxFreqOffsetHz, freedv_get_modem_sample_rate(dv), st->rxFreqOffsetPhaseRectObj, nin);
+                nout = freedv_comprx(dv, output_buf, rx_fdm_offset);
             
-            // Resample output to the current mode's rate if needed.
-            nout = resample(soundOutRateConv_, output_resample_buf, output_buf, getRxSpeechSampleRate(), freedv_get_speech_sample_rate(dv), getRxNumSpeechSamples(), nout);
-
-            // Write to output FIFO on one of the following conditions:
-            //   a) We're on the last mode to check (meaning that we didn't get sync on any other mode)
-            //.  b) We got sync on the current mode at some point
-            int tmpState = freedv_get_sync(dv);
-            if (tmpState != 0 || done || (size_t)index == (dvObjects_.size() - 1))
+                // Resample output to the current mode's rate if needed.
+                {
+                    std::lock_guard<std::mutex> lock(resampleMtx_);
+                    nout = resample(st->soundOutRateConv, output_resample_buf, output_buf, st->rxSpeechSampleRate, freedv_get_speech_sample_rate(dv), numSpeechSamples, nout);
+                }
+                
+                // Write to output FIFO if we have sync.
+                int tmpState = freedv_get_sync(dv);
+                if (tmpState != 0)
+                {
+                    codec2_fifo_write(st->ownOutput, output_resample_buf, nout);
+                }
+            
+                nin = freedv_nin(dv);
+            }
+                    
+            return st;
+        }, inpState));
+    }
+    
+    // Loop through each future and wait for the result, then determine
+    // which mode is most likely to have sync. Default to the last mode
+    // in the list if we don't find any.
+    int futIndexWithSync = rxFutures.size() - 1;
+    int futIndex = 0;
+    int maxSyncFound = 0;
+    std::deque<RxAudioThreadState*> results;
+    for(auto& fut : rxFutures)
+    {
+        fut.wait();
+        RxAudioThreadState* res = fut.get();
+        results.push_back(res);
+        
+        struct MODEM_STATS tmpStats;
+        freedv_get_modem_extended_stats(res->dvObj, &tmpStats);
+        
+        int snr = -10;
+        if (!isnan(tmpStats.snr_est) && !isinf(tmpStats.snr_est))
+            snr = tmpStats.snr_est + 0.5;
+        
+        if (lastSyncRxMode_ == res->dvObj)
+        {
+            if (stats->sync != 0)
             {
-                state = tmpState;
-                done = true;
-                currentRxMode_ = dv;
-                rxMode_ = enabledModes_[index];
-                codec2_fifo_write(outputFifo, output_resample_buf, nout);
+                futIndexWithSync = futIndex;
+                break;
             }
             
-            nin = freedv_nin(dv);
+            lastSyncRxMode_ = nullptr;
         }
         
-        if (done)
+        if (snr > maxSyncFound && tmpStats.sync != 0)
```

Comment:
@drowe67, using both sync and highest SNR _should_ be okay if 800XA and 2400B aren't included in multi-RX. Based on my understanding, if there are somehow two or more modes that are in sync (due to blips), the highest SNR should be the "correct" one. This could be a problem for signals that are within the noise but there's also have code farther up to not select any other modes if we're already in sync on one of them.

Anyway, let me know if my understanding is incorrect or if there are additional pitfalls I didn't think of. 👍

Reply:
Its a bit hard to say, as the SNR is filtered/averaged so may rise slower in some modes than others.   However if a mistake is made it will fall out of sync eventually and you'll get another chance.